### ISSUE TITLE
Tuning of Yolov8 :(Memory occupied by different train folder's after tuning. )

### ISSUE THREAD

**Issue by DANISHFAYAZNAJAR** [ultralytics/ultralytics#7123]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 

I was tuning Yolo for hyper parameter's so, I tried 

`
model = YOLO('yolov8l.pt')

model.tune(epochs=10,iterations=50,
           data=dataset,
           project=project_name,
           optimizer='Adam',
           batch=32)`
           
When I checked out the memory occupied by each train folder it's different. Why is it so?. As we are using same model for each run, so same size model.

Or it's because each model is trained with different hyper parameters, so model learn differently, so different weights, and the space occupied by each weight varies and in total space occupied by different model is different.

Like if some weights are 0 so, it requires less bit's to store and if some weights are high precision then it requires more bit's to store.

I feel this can be a reason, but want to confirm from experts.
<img width="180" alt="Screenshot 2023-12-22 at 11 56 33 AM" src="https://github.com/ultralytics/ultralytics/assets/70327223/6b0fcc9a-ec90-4ac3-a840-f279e64057bf">

**Comment by DANISHFAYAZNAJAR** [comment]
Also after tuning should we choose best_fitness model or model with highest required metrics. Like mAP50 highest model or mAP50-95.

**Comment by DANISHFAYAZNAJAR** [comment]
As after model tune we get a tune folder, in which we get best_hyperparameters.yaml file. Following is content from best_hyperparameters.yaml file.

`
# Best fitness=0.68464 observed at iteration 38
# Best fitness metrics are {'metrics/precision(B)': 0.88868, 'metrics/recall(B)': 0.89105, 'metrics/mAP50(B)': 0.92822, 'metrics/mAP50-95(B)': 0.65757, 'val/box_loss': 1.03365, 'val/cls_loss': 0.77574, 'val/dfl_loss': 1.48296, 'fitness': 0.68464}
# Best fitness model is project/train38`


But in logs  of train38 I find below logs:

`
 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  70%|███████   | 7/10 [00:36<00:10,  3.42s/it]
74                    all        587      12320      0.884      0.777      0.814      0.576
75                 class1        587       6882      0.907      0.785       0.81      0.563
76                 class2        587       1536      0.793      0.712      0.751      0.445
77                    class3        587       3317      0.858      0.755      0.823      0.481
78                  class4        587        585       0.98      0.855      0.873      0.814
79 Speed: 0.5ms preprocess, 7.3ms inference, 0.0ms loss, 14.0ms postprocess per image
80 Results saved to project/train38
`

So the metrics in logs and in best_hyperparameters doesn't match.

Therefore why tune has kept best_fitness model as train38, when in logs of train38 we can see the metrics are low. Can this be because of Broken Pipe, as my tune stopped at iterations 39, so we got the last train as best one?

**Comment by glenn-jocher** [comment]
@DANISHFAYAZNAJAR hello!

Regarding the varying memory occupancy of each train folder after tuning, it's normal to observe some differences. This can be due to several factors, including variations in the learned weights, differences in optimizer state, logging information, and more. Even when starting from the same initial model, the tuning process leads to unique outcomes due to the exploration of the hyperparameter space. Each run can lead to slightly different weight values, which may affect storage size, especially when considering precision and the way floating-point numbers are stored.

When selecting the final model after tuning, it's commonly suggested to prioritize the model with the highest 'fitness' metric determined by the tuning process, as that considers a balance of precision, recall, and mAP scores. If you have a specific requirement or you value one metric significantly more than others, you could select the model with the highest value of that particular metric – for example, the highest mAP50 or mAP50-95. Ultimately, the choice depends on your application and the performance characteristics you prioritize most.

For your second question about the discrepancy between the `best_hyperparameters.yaml` content and the logs, the 'best' designation is generally based on overall performance across the total tuning runs. However, if the process stopped prematurely due to an issue (like a Broken Pipe or any interruption), the evaluation might default to the last or the best-seen configuration up to that point. It's advisable to ensure uninterrupted tuning runs for the most accurate assessment of best performance. If you encounter discrepancies, it might be beneficial to rerun the tuning if possible to ensure the results are robust and reliable.

Happy model tuning! 🚀

**Comment by DANISHFAYAZNAJAR** [comment]
@glenn-jocher  Thanks for clearing my doubt.

**Comment by glenn-jocher** [comment]
@DANISHFAYAZNAJAR you're welcome! If you have any more questions or need further assistance, feel free to ask. Happy to help! 😊

**Comment by DANISHFAYAZNAJAR** [comment]
@glenn-jocher  Also I found that only few of the run's among tuning contain's weights. Is it because of tuning failed, it saved it in few train's only or it's default.

**Comment by glenn-jocher** [comment]
@DANISHFAYAZNAJAR during tuning, weights are saved selectively to manage resources efficiently. This is by design and not indicative of a tuning failure. Weights may be saved for specific runs that meet certain criteria, such as improved metrics or at designated intervals. This ensures the most promising models are retained without overwhelming storage with every iteration's weights. If you're seeing weights only for certain runs, it's likely behaving as intended. 👍
